File: skills/openclix-init/templates/flutter/engine/campaign_processor.dart
import '../models/clix_types.dart';
import '../services/utils.dart';
import 'event_condition_processor.dart';
import 'schedule_calculator.dart';

class CampaignDecision {
  final DecisionTrace trace;
  final String action;
  final QueuedMessage? queuedMessage;
  final String? scheduledFor;

  CampaignDecision({
    required this.trace,
    required this.action,
    this.queuedMessage,
    this.scheduledFor,
  });
}

class CampaignProcessorDependencies {
  final EventConditionProcessor eventConditionProcessor;
  final ScheduleCalculator scheduleCalculator;
  final ClixLogger logger;
  final Settings? settings;

  CampaignProcessorDependencies({
    required this.eventConditionProcessor,
    required this.scheduleCalculator,
    required this.logger,
    this.settings,
  });
}

DecisionTrace createTrace(
  String campaignId,
  String action,
  String result,
  String reason, [
  SkipReason? skipReason,
]) {
  return DecisionTrace(
    campaignId: campaignId,
    action: action,
    result: result,
    skipReason: skipReason,
    reason: reason,
  );
}

CampaignDecision createSkipDecision(
  String campaignId,
  String reason, [
  SkipReason? skipReason,
]) {
  return CampaignDecision(
    action: 'skip',
    trace: createTrace(
      campaignId,
      'skip_campaign',
      'skipped',
      reason,
      skipReason,
    ),
  );
}

class ExecutionResolution {
  final String? executeAt;
  final String? scheduledFor;
  final String? triggerEventId;
  final String? reason;
  final SkipReason? skipReason;

  ExecutionResolution({
    this.executeAt,
    this.scheduledFor,
    this.triggerEventId,
    this.reason,
    this.skipReason,
  });
}

final Map<DayOfWeek, int> dayIndex = {
  DayOfWeek.sunday: 0,
  DayOfWeek.monday: 1,
  DayOfWeek.tuesday: 2,
  DayOfWeek.wednesday: 3,
  DayOfWeek.thursday: 4,
  DayOfWeek.friday: 5,
  DayOfWeek.saturday: 6,
};

DateTime? parseDateTimeMaybe(String? value) {
  if (value == null || value.isEmpty) return null;
  try {
    return DateTime.parse(value).toUtc();
  } catch (_) {
    return null;
  }
}

DateTime startOfWeek(DateTime dateTime) {
  final normalized = DateTime.utc(dateTime.year, dateTime.month, dateTime.day);
  return normalized.subtract(Duration(days: normalized.weekday % 7));
}

DateTime withTime(DateTime dateTime, int hour, int minute) {
  return DateTime.utc(
    dateTime.year,
    dateTime.month,
    dateTime.day,
    hour,
    minute,
  );
}

class CampaignProcessor {
  CampaignDecision process(
    String campaignId,
    Campaign campaign,
    TriggerContext context,
    CampaignStateSnapshot snapshot,
    CampaignProcessorDependencies dependencies,
  ) {
    final eventConditionProcessor = dependencies.eventConditionProcessor;
    final scheduleCalculator = dependencies.scheduleCalculator;
    final logger = dependencies.logger;
    final settings = dependencies.settings;

    final now = context.now ?? DateTime.now().toUtc().toIso8601String();
    final campaignState = getCampaignState(snapshot, campaignId);

    if (campaign.status != CampaignStatus.running) {
      return createSkipDecision(
        campaignId,
        "Campaign status is '${campaign.status.value}', not 'running'",
        SkipReason.campaignNotRunning,
      );
    }

    if ((campaign.trigger.type == TriggerType.event &&
            context.trigger != 'event_tracked') ||
        (campaign.trigger.type != TriggerType.event &&
            context.trigger == 'event_tracked')) {
      return createSkipDecision(
        campaignId,
        "Trigger type '${campaign.trigger.type.value}' is not eligible for '${context.trigger}'",
      );
    }

    if (campaign.trigger.type != TriggerType.recurring &&
        campaignState?.triggered == true) {
      return createSkipDecision(campaignId, 'Campaign already triggered');
    }

    if (settings?.frequencyCap != null) {
      final maxCount = settings!.frequencyCap!.maxCount;
      final windowSeconds = settings.frequencyCap!.windowSeconds;
      final nowDate = parseDateTimeMaybe(now) ?? DateTime.now().toUtc();
      final windowStart = nowDate
          .subtract(Duration(seconds: windowSeconds))
          .toIso8601String();

      final recentTriggerCount = snapshot.triggerHistory
          .where((history) => history.triggeredAt.compareTo(windowStart) >= 0)
          .length;

      if (recentTriggerCount >= maxCount) {
        return createSkipDecision(
          campaignId,
          'Frequency cap exceeded '
          '($recentTriggerCount/$maxCount within ${windowSeconds}s)',
          SkipReason.campaignFrequencyCapExceeded,
        );
      }
    }

    if (campaign.trigger.type == TriggerType.recurring &&
        hasFuturePendingForCampaign(snapshot, campaignId, now)) {
      return createSkipDecision(
        campaignId,
        'Recurring campaign already has a queued message',
      );
    }

    final resolved = resolveExecutionTime(
      campaign,
      context,
      campaignState,
      now,
      eventConditionProcessor,
    );

    if (resolved.executeAt == null) {
      return createSkipDecision(
        campaignId,
        resolved.reason ?? 'Campaign trigger conditions were not met',
        resolved.skipReason,
      );
    }

    if (hasPendingForCampaignAt(snapshot, campaignId, resolved.executeAt!)) {
      return createSkipDecision(
        campaignId,
        'Duplicate schedule prevented for campaign at ${resolved.executeAt}',
      );
    }

    final scheduleResult = scheduleCalculator.calculate(
      ScheduleInput(
        now: now,
        executeAt: resolved.executeAt,
        doNotDisturb: settings?.doNotDisturb,
      ),
    );

    if (scheduleResult.skipped) {
      return createSkipDecision(
        campaignId,
        'Blocked by do-not-disturb window',
        scheduleResult.skipReason,
      );
    }

    final templateVariables = <String, dynamic>{
      ...(context.event?.properties ?? const {}),
    };

    final renderedTitle = renderTemplate(
      campaign.message.content.title,
      templateVariables,
    );
    final renderedBody = renderTemplate(
      campaign.message.content.body,
      templateVariables,
    );

    final queuedMessage = QueuedMessage(
      id: generateUUID(),
      campaignId: campaignId,
      channelType: campaign.message.channelType,
      status: QueuedMessageStatus.scheduled,
      content: MessageContent(
        title: renderedTitle,
        body: renderedBody,
        imageUrl: campaign.message.content.imageUrl,
        landingUrl: campaign.message.content.landingUrl,
      ),
      triggerEventId: resolved.triggerEventId,
      executeAt: scheduleResult.executeAt,
      createdAt: now,
    );

    logger.debug(
      '[CampaignProcessor] Campaign $campaignId: triggered, scheduled for ${scheduleResult.executeAt}',
    );

    return CampaignDecision(
      action: 'trigger',
      trace: createTrace(
        campaignId,
        'trigger_campaign',
        'applied',
        'Campaign triggered, message scheduled for ${scheduleResult.executeAt}',
      ),
      queuedMessage: queuedMessage,
      scheduledFor: resolved.scheduledFor ?? scheduleResult.executeAt,
    );
  }

  ExecutionResolution resolveExecutionTime(
    Campaign campaign,
    TriggerContext context,
    CampaignStateRecord? campaignState,
    String now,
    EventConditionProcessor eventConditionProcessor,
  ) {
    if (campaign.trigger.type == TriggerType.event) {
      final eventConfiguration = campaign.trigger.event;
      if (eventConfiguration == null) {
        return ExecutionResolution(
          reason: "Trigger type 'event' requires trigger.event configuration",
          skipReason: SkipReason.triggerEventNotMatched,
        );
      }

      if (context.event == null) {
        return ExecutionResolution(
          reason: 'Event trigger requires an event in context',
          skipReason: SkipReason.triggerEventNotMatched,
        );
      }

      final matched = eventConditionProcessor.process(
        eventConfiguration.triggerEvent,
        context.event!,
      );
      if (!matched) {
        return ExecutionResolution(
          reason:
              "Trigger event conditions did not match event '${context.event!.name}'",
          skipReason: SkipReason.triggerEventNotMatched,
        );
      }

      final delaySeconds = eventConfiguration.delaySeconds ?? 0;
      final nowDate = parseDateTimeMaybe(now) ?? DateTime.now().toUtc();
      final executeAt = nowDate
          .add(Duration(seconds: delaySeconds))
          .toIso8601String();

      return ExecutionResolution(
        executeAt: executeAt,
        scheduledFor: executeAt,
        triggerEventId: context.event!.id,
      );
    }

    if (campaign.trigger.type == TriggerType.scheduled) {
      final scheduledConfiguration = campaign.trigger.scheduled;
      final executeAtDate = parseDateTimeMaybe(
        scheduledConfiguration?.executeAt,
      );
      if (scheduledConfiguration == null || executeAtDate == null) {
        return ExecutionResolution(
          reason: 'Scheduled trigger requires a valid execute_at datetime',
        );
      }

      final nowDate = parseDateTimeMaybe(now) ?? DateTime.now().toUtc();
      if (!executeAtDate.isAfter(nowDate)) {
        return ExecutionResolution(
          reason:
              "Scheduled execute_at '${scheduledConfiguration.executeAt}' is already in the past",
        );
      }

      return ExecutionResolution(
        executeAt: scheduledConfiguration.executeAt,
        scheduledFor: scheduledConfiguration.executeAt,
      );
    }

    final recurringConfiguration = campaign.trigger.recurring;
    if (recurringConfiguration == null) {
      return ExecutionResolution(
        reason: 'Recurring trigger requires recurring configuration',
      );
    }

    final nextExecuteAt = computeNextRecurringExecuteAt(
      recurringConfiguration,
      now,
      campaignState?.recurringLastScheduledAt,
      campaignState?.recurringAnchorAt,
    );

    if (nextExecuteAt == null) {
      return ExecutionResolution(
        reason: 'Recurring schedule has no upcoming execution window',
      );
    }

    return ExecutionResolution(
      executeAt: nextExecuteAt,
      scheduledFor: nextExecuteAt,
    );
  }

  String? computeNextRecurringExecuteAt(
    RecurringTriggerConfig recurringConfiguration,
    String now,
    String? lastScheduledAt,
    String? recurringAnchorAt,
  ) {
    final nowDate = parseDateTimeMaybe(now);
    if (nowDate == null) return null;

    final startAtDate = parseDateTimeMaybe(recurringConfiguration.startAt);
    final recurringAnchorDate = parseDateTimeMaybe(recurringAnchorAt);
    final startDate =
        startAtDate ??
        recurringAnchorDate ??
        computeDefaultRecurringAnchorDate(recurringConfiguration, nowDate);

    final endDate = recurringConfiguration.endAt == null
        ? null
        : parseDateTimeMaybe(recurringConfiguration.endAt);
    if (recurringConfiguration.endAt != null && endDate == null) {
      return null;
    }

    final fromDate =
        parseDateTimeMaybe(lastScheduledAt)?.add(const Duration(minutes: 1)) ??
        nowDate;

    final interval = recurringConfiguration.rule.interval < 1
        ? 1
        : recurringConfiguration.rule.interval;

    final next = findNextOccurrence(
      recurringConfiguration,
      startDate,
      fromDate,
      interval,
    );

    if (next == null) return null;
    if (endDate != null && next.isAfter(endDate)) return null;

    return next.toIso8601String();
  }

  DateTime computeDefaultRecurringAnchorDate(
    RecurringTriggerConfig recurringConfiguration,
    DateTime nowDate,
  ) {
    if (recurringConfiguration.rule.type == RecurrenceType.hourly) {
      return DateTime.utc(
        nowDate.year,
        nowDate.month,
        nowDate.day,
        nowDate.hour,
      );
    }

    if (recurringConfiguration.rule.type == RecurrenceType.daily) {
      final timeOfDay = recurringConfiguration.rule.timeOfDay;
      final hour = timeOfDay?.hour ?? nowDate.hour;
      final minute = timeOfDay?.minute ?? nowDate.minute;
      return DateTime.utc(
        nowDate.year,
        nowDate.month,
        nowDate.day,
        hour,
        minute,
      );
    }

    final timeOfDay = recurringConfiguration.rule.timeOfDay;
    final hour = timeOfDay?.hour ?? nowDate.hour;
    final minute = timeOfDay?.minute ?? nowDate.minute;
    final weekAnchor = startOfWeek(nowDate);
    return DateTime.utc(
      weekAnchor.year,
      weekAnchor.month,
      weekAnchor.day,
      hour,
      minute,
    );
  }

  DateTime? findNextOccurrence(
    RecurringTriggerConfig recurringConfiguration,
    DateTime startDate,
    DateTime fromDate,
    int interval,
  ) {
    final rule = recurringConfiguration.rule;

    if (rule.type == RecurrenceType.hourly) {
      final baseMilliseconds = startDate.millisecondsSinceEpoch;
      final fromMilliseconds =
          fromDate.millisecondsSinceEpoch > baseMilliseconds
          ? fromDate.millisecondsSinceEpoch
          : baseMilliseconds;
      final intervalMilliseconds = interval * 60 * 60 * 1000;
      final steps =
          ((fromMilliseconds - baseMilliseconds) / intervalMilliseconds).ceil();
      return DateTime.fromMillisecondsSinceEpoch(
        baseMilliseconds + (steps < 0 ? 0 : steps) * intervalMilliseconds,
        isUtc: true,
      );
    }

    if (rule.type == RecurrenceType.daily) {
      final hour = rule.timeOfDay?.hour ?? startDate.hour;
      final minute = rule.timeOfDay?.minute ?? startDate.minute;

      var candidate = withTime(startDate, hour, minute);
      if (candidate.isBefore(fromDate)) {
        final deltaDays = fromDate.difference(candidate).inDays;
        final steps = (deltaDays / interval).floor();
        candidate = candidate.add(Duration(days: steps * interval));

        while (candidate.isBefore(fromDate)) {
          candidate = candidate.add(Duration(days: interval));
        }
      }

      return candidate;
    }

    if (rule.type == RecurrenceType.weekly) {
      final days = rule.weeklyRule?.daysOfWeek ?? const <DayOfWeek>[];
      if (days.isEmpty) return null;

      final allowedDays = days.map((day) => dayIndex[day]!).toSet();
      final hour = rule.timeOfDay?.hour ?? startDate.hour;
      final minute = rule.timeOfDay?.minute ?? startDate.minute;
      final anchorWeekStart = startOfWeek(startDate).millisecondsSinceEpoch;

      final cursor = DateTime.utc(fromDate.year, fromDate.month, fromDate.day);

      for (var offset = 0; offset < 732; offset += 1) {
        final candidateDay = cursor.add(Duration(days: offset));
        final candidate = DateTime.utc(
          candidateDay.year,
          candidateDay.month,
          candidateDay.day,
          hour,
          minute,
        );

        if (candidate.isBefore(fromDate)) continue;

        final candidateDayIndex = candidate.weekday % 7;
        if (!allowedDays.contains(candidateDayIndex)) continue;

        final candidateWeekStart = startOfWeek(
          candidate,
        ).millisecondsSinceEpoch;
        final weekDiff =
            ((candidateWeekStart - anchorWeekStart) / (7 * 24 * 60 * 60 * 1000))
                .floor();

        if (weekDiff >= 0 && weekDiff % interval == 0) {
          return candidate;
        }
      }

      return null;
    }

    return null;
  }

  bool hasPendingForCampaignAt(
    CampaignStateSnapshot snapshot,
    String campaignId,
    String executeAt,
  ) {
    return snapshot.queuedMessages.any(
      (pending) =>
          pending.campaignId == campaignId && pending.executeAt == executeAt,
    );
  }

  bool hasFuturePendingForCampaign(
    CampaignStateSnapshot snapshot,
    String campaignId,
    String now,
  ) {
    final nowDate = parseDateTimeMaybe(now);

    for (final pending in snapshot.queuedMessages) {
      if (pending.campaignId != campaignId) continue;

      final executeAtDate = parseDateTimeMaybe(pending.executeAt);
      if (executeAtDate == null) return true;
      if (nowDate == null || !executeAtDate.isBefore(nowDate)) {
        return true;
      }
    }

    return false;
  }

  CampaignStateRecord? getCampaignState(
    CampaignStateSnapshot snapshot,
    String campaignId,
  ) {
    for (final state in snapshot.campaignStates) {
      if (state.campaignId == campaignId) {
        return state;
      }
    }
    return null;
  }
}

File: skills/openclix-init/templates/ios/Engine/CampaignProcessor.swift
import Foundation

public struct CampaignDecision {
    public let trace: DecisionTrace
    public let action: CampaignAction
    public let queued_message: QueuedMessage?
    public let scheduled_for: String?

    public init(
        trace: DecisionTrace,
        action: CampaignAction,
        queued_message: QueuedMessage? = nil,
        scheduled_for: String? = nil
    ) {
        self.trace = trace
        self.action = action
        self.queued_message = queued_message
        self.scheduled_for = scheduled_for
    }
}

public enum CampaignAction {
    case trigger
    case skip
}

public struct CampaignProcessorDependencies {
    public let eventConditionProcessor: EventConditionProcessor
    public let scheduleCalculator: ScheduleCalculator
    public let logger: ClixLogger
    public let settings: Settings?

    public init(
        eventConditionProcessor: EventConditionProcessor,
        scheduleCalculator: ScheduleCalculator,
        logger: ClixLogger,
        settings: Settings? = nil
    ) {
        self.eventConditionProcessor = eventConditionProcessor
        self.scheduleCalculator = scheduleCalculator
        self.logger = logger
        self.settings = settings
    }
}

private func createTrace(
    campaign_id: String,
    action: String,
    result: DecisionResult,
    reason: String,
    skip_reason: SkipReason? = nil
) -> DecisionTrace {
    return DecisionTrace(
        campaign_id: campaign_id,
        action: action,
        result: result,
        skip_reason: skip_reason,
        reason: reason
    )
}

private func createSkipDecision(
    campaignId: String,
    reason: String,
    skipReason: SkipReason? = nil
) -> CampaignDecision {
    return CampaignDecision(
        trace: createTrace(
            campaign_id: campaignId,
            action: "skip_campaign",
            result: .skipped,
            reason: reason,
            skip_reason: skipReason
        ),
        action: .skip
    )
}

private struct ExecutionResolution {
    let executeAt: String?
    let scheduledFor: String?
    let triggerEventId: String?
    let reason: String?
    let skipReason: SkipReason?
}

private let dayIndex: [DayOfWeek: Int] = [
    .sunday: 0,
    .monday: 1,
    .tuesday: 2,
    .wednesday: 3,
    .thursday: 4,
    .friday: 5,
    .saturday: 6,
]

private func parseIsoDate(_ value: String?) -> Date? {
    guard let value = value, !value.isEmpty else { return nil }

    let internetFormatter = ISO8601DateFormatter()
    internetFormatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
    if let parsed = internetFormatter.date(from: value) {
        return parsed
    }

    let fallbackFormatter = ISO8601DateFormatter()
    return fallbackFormatter.date(from: value)
}

private func toIsoString(_ date: Date) -> String {
    let formatter = ISO8601DateFormatter()
    formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
    return formatter.string(from: date)
}

private func startOfWeek(_ date: Date) -> Date {
    var calendar = Calendar.current
    calendar.timeZone = .current

    let weekday = calendar.component(.weekday, from: date)
    let daysSinceSunday = weekday - 1
    let midnight = calendar.startOfDay(for: date)
    return calendar.date(byAdding: .day, value: -daysSinceSunday, to: midnight) ?? midnight
}

private func withTime(_ base: Date, hour: Int, minute: Int) -> Date {
    var calendar = Calendar.current
    calendar.timeZone = .current

    var components = calendar.dateComponents([.year, .month, .day], from: base)
    components.hour = hour
    components.minute = minute
    components.second = 0
    components.nanosecond = 0
    return calendar.date(from: components) ?? base
}

private func decodeJsonValue(_ value: JsonValue) -> Any {
    switch value {
    case .string(let text):
        return text
    case .number(let number):
        return number
    case .bool(let value):
        return value
    case .null:
        return NSNull()
    case .array(let values):
        return values.map { decodeJsonValue($0) }
    case .object(let values):
        return values.mapValues { decodeJsonValue($0) }
    }
}

public final class CampaignProcessor {

    public init() {}

    public func process(
        campaignId: String,
        campaign: Campaign,
        context: TriggerContext,
        snapshot: CampaignStateSnapshot,
        dependencies: CampaignProcessorDependencies
    ) -> CampaignDecision {
        let eventConditionProcessor = dependencies.eventConditionProcessor
        let scheduleCalculator = dependencies.scheduleCalculator
        let logger = dependencies.logger
        let settings = dependencies.settings
        let now = context.now ?? toIsoString(Date())
        let nowDate = parseIsoDate(now) ?? Date()

        let campaignState = getCampaignState(snapshot: snapshot, campaignId: campaignId)

        if campaign.status != .running {
            return createSkipDecision(
                campaignId: campaignId,
                reason: "Campaign status is '\(campaign.status.rawValue)', not 'running'",
                skipReason: .campaign_not_running
            )
        }

        if (campaign.trigger.type == .event && context.trigger != .event_tracked)
            || (campaign.trigger.type != .event && context.trigger == .event_tracked) {
            return createSkipDecision(
                campaignId: campaignId,
                reason: "Trigger type '\(campaign.trigger.type.rawValue)' is not eligible for '\(context.trigger.rawValue)'"
            )
        }

        if campaign.trigger.type != .recurring,
           campaignState?.triggered == true {
            return createSkipDecision(
                campaignId: campaignId,
                reason: "Campaign already triggered"
            )
        }

        if let frequencyCap = settings?.frequency_cap {
            let windowStartDate = nowDate.addingTimeInterval(-Double(frequencyCap.window_seconds))
            let countInWindow = snapshot.trigger_history.reduce(into: 0) { partial, row in
                guard let triggeredAtDate = parseIsoDate(row.triggered_at) else { return }
                if triggeredAtDate >= windowStartDate {
                    partial += 1
                }
            }

            if countInWindow >= frequencyCap.max_count {
                return createSkipDecision(
                    campaignId: campaignId,
                    reason: "Frequency cap exceeded (\(countInWindow)/\(frequencyCap.max_count) within \(frequencyCap.window_seconds)s)",
                    skipReason: .campaign_frequency_cap_exceeded
                )
            }
        }

        if campaign.trigger.type == .recurring,
           hasFuturePendingForCampaign(
                snapshot: snapshot,
                campaignId: campaignId,
                now: now
           ) {
            return createSkipDecision(
                campaignId: campaignId,
                reason: "Recurring campaign already has a queued message"
            )
        }

        let resolved = resolveExecutionTime(
            campaign: campaign,
            context: context,
            campaignState: campaignState,
            now: now,
            nowDate: nowDate,
            eventConditionProcessor: eventConditionProcessor
        )

        guard let executeAt = resolved.executeAt else {
            return createSkipDecision(
                campaignId: campaignId,
                reason: resolved.reason ?? "Campaign trigger conditions were not met",
                skipReason: resolved.skipReason
            )
        }

        if hasPendingForCampaignAt(
            snapshot: snapshot,
            campaignId: campaignId,
            executeAt: executeAt
        ) {
            return createSkipDecision(
                campaignId: campaignId,
                reason: "Duplicate schedule prevented for campaign at \(executeAt)"
            )
        }

        let scheduleResult = scheduleCalculator.calculate(
            ScheduleInput(
                now: now,
                execute_at: executeAt,
                do_not_disturb: settings?.do_not_disturb
            )
        )

        if scheduleResult.skipped {
            return createSkipDecision(
                campaignId: campaignId,
                reason: "Blocked by do-not-disturb window",
                skipReason: scheduleResult.skip_reason
            )
        }

        var templateVariables: [String: Any] = [:]
        if let eventProperties = context.event?.properties {
            for (key, value) in eventProperties {
                templateVariables[key] = decodeJsonValue(value)
            }
        }

        let renderedTitle = renderTemplate(campaign.message.content.title, variables: templateVariables)
        let renderedBody = renderTemplate(campaign.message.content.body, variables: templateVariables)

        let queuedMessage = QueuedMessage(
            id: generateUUID(),
            campaign_id: campaignId,
            channel_type: campaign.message.channel_type,
            status: .scheduled,
            content: QueuedMessageContent(
                title: renderedTitle,
                body: renderedBody,
                image_url: campaign.message.content.image_url,
                landing_url: campaign.message.content.landing_url
            ),
            trigger_event_id: resolved.triggerEventId,
            execute_at: scheduleResult.execute_at,
            created_at: now
        )

        logger.debug(
            "[CampaignProcessor] Campaign \(campaignId): triggered, scheduled for \(scheduleResult.execute_at)"
        )

        return CampaignDecision(
            trace: createTrace(
                campaign_id: campaignId,
                action: "trigger_campaign",
                result: .applied,
                reason: "Campaign triggered, message scheduled for \(scheduleResult.execute_at)"
            ),
            action: .trigger,
            queued_message: queuedMessage,
            scheduled_for: resolved.scheduledFor ?? scheduleResult.execute_at
        )
    }

    private func resolveExecutionTime(
        campaign: Campaign,
        context: TriggerContext,
        campaignState: CampaignStateRecord?,
        now: String,
        nowDate: Date,
        eventConditionProcessor: EventConditionProcessor
    ) -> ExecutionResolution {
        if campaign.trigger.type == .event {
            guard let eventConfig = campaign.trigger.event else {
                return ExecutionResolution(
                    executeAt: nil,
                    scheduledFor: nil,
                    triggerEventId: nil,
                    reason: "Trigger type 'event' requires trigger.event configuration",
                    skipReason: .trigger_event_not_matched
                )
            }

            guard let event = context.event else {
                return ExecutionResolution(
                    executeAt: nil,
                    scheduledFor: nil,
                    triggerEventId: nil,
                    reason: "Event trigger requires an event in context",
                    skipReason: .trigger_event_not_matched
                )
            }

            let matched = eventConditionProcessor.process(
                group: eventConfig.trigger_event,
                event: event
            )

            if !matched {
                return ExecutionResolution(
                    executeAt: nil,
                    scheduledFor: nil,
                    triggerEventId: nil,
                    reason: "Trigger event conditions did not match event '\(event.name)'",
                    skipReason: .trigger_event_not_matched
                )
            }

            let delaySeconds = eventConfig.delay_seconds ?? 0
            let executeAtDate = nowDate.addingTimeInterval(Double(delaySeconds))
            let executeAt = toIsoString(executeAtDate)

            return ExecutionResolution(
                executeAt: executeAt,
                scheduledFor: executeAt,
                triggerEventId: event.id,
                reason: nil,
                skipReason: nil
            )
        }

        if campaign.trigger.type == .scheduled {
            guard let scheduled = campaign.trigger.scheduled,
                  let executeAtDate = parseIsoDate(scheduled.execute_at) else {
                return ExecutionResolution(
                    executeAt: nil,
                    scheduledFor: nil,
                    triggerEventId: nil,
                    reason: "Scheduled trigger requires a valid execute_at datetime",
                    skipReason: nil
                )
            }

            if executeAtDate <= nowDate {
                return ExecutionResolution(
                    executeAt: nil,
                    scheduledFor: nil,
                    triggerEventId: nil,
                    reason: "Scheduled execute_at '\(scheduled.execute_at)' is already in the past",
                    skipReason: nil
                )
            }

            return ExecutionResolution(
                executeAt: scheduled.execute_at,
                scheduledFor: scheduled.execute_at,
                triggerEventId: nil,
                reason: nil,
                skipReason: nil
            )
        }

        guard let recurring = campaign.trigger.recurring else {
            return ExecutionResolution(
                executeAt: nil,
                scheduledFor: nil,
                triggerEventId: nil,
                reason: "Recurring trigger requires recurring configuration",
                skipReason: nil
            )
        }

        let nextExecuteAt = computeNextRecurringExecuteAt(
            recurring: recurring,
            now: now,
            lastScheduledAt: campaignState?.recurring_last_scheduled_at,
            recurringAnchorAt: campaignState?.recurring_anchor_at
        )

        guard let nextExecuteAt else {
            return ExecutionResolution(
                executeAt: nil,
                scheduledFor: nil,
                triggerEventId: nil,
                reason: "Recurring schedule has no upcoming execution window",
                skipReason: nil
            )
        }

        return ExecutionResolution(
            executeAt: nextExecuteAt,
            scheduledFor: nextExecuteAt,
            triggerEventId: nil,
            reason: nil,
            skipReason: nil
        )
    }

    private func computeNextRecurringExecuteAt(
        recurring: RecurringTriggerConfig,
        now: String,
        lastScheduledAt: String?,
        recurringAnchorAt: String?
    ) -> String? {
        guard let nowDate = parseIsoDate(now) else { return nil }

        let startDate: Date
        if let recurringStartDate = parseIsoDate(recurring.start_at) {
            startDate = recurringStartDate
        } else if let anchorDate = parseIsoDate(recurringAnchorAt) {
            startDate = anchorDate
        } else {
            startDate = computeDefaultRecurringAnchorDate(recurring: recurring, nowDate: nowDate)
        }

        let endDate: Date?
        if recurring.end_at != nil {
            endDate = parseIsoDate(recurring.end_at)
            if endDate == nil {
                return nil
            }
        } else {
            endDate = nil
        }

        let fromDate: Date
        if let lastScheduledDate = parseIsoDate(lastScheduledAt) {
            fromDate = lastScheduledDate.addingTimeInterval(60)
        } else {
            fromDate = nowDate
        }

        let interval = max(1, recurring.rule.interval)
        guard let nextOccurrence = findNextOccurrence(
            recurring: recurring,
            startDate: startDate,
            fromDate: fromDate,
            interval: interval
        ) else {
            return nil
        }

        if let endDate, nextOccurrence > endDate {
            return nil
        }

        return toIsoString(nextOccurrence)
    }

    private func computeDefaultRecurringAnchorDate(
        recurring: RecurringTriggerConfig,
        nowDate: Date
    ) -> Date {
        var calendar = Calendar.current
        calendar.timeZone = .current

        var anchor = nowDate
        if recurring.rule.type == .hourly {
            anchor = calendar.date(bySettingHour: calendar.component(.hour, from: nowDate), minute: 0, second: 0, of: nowDate) ?? nowDate
            return anchor
        }

        if recurring.rule.type == .daily {
            let hour = recurring.rule.time_of_day?.hour ?? calendar.component(.hour, from: nowDate)
            let minute = recurring.rule.time_of_day?.minute ?? calendar.component(.minute, from: nowDate)
            anchor = calendar.date(bySettingHour: hour, minute: minute, second: 0, of: nowDate) ?? nowDate
            return anchor
        }

        let hour = recurring.rule.time_of_day?.hour ?? calendar.component(.hour, from: nowDate)
        let minute = recurring.rule.time_of_day?.minute ?? calendar.component(.minute, from: nowDate)
        let weekAnchor = startOfWeek(nowDate)
        return withTime(weekAnchor, hour: hour, minute: minute)
    }

    private func findNextOccurrence(
        recurring: RecurringTriggerConfig,
        startDate: Date,
        fromDate: Date,
        interval: Int
    ) -> Date? {
        let daySeconds: TimeInterval = 24 * 60 * 60

        if recurring.rule.type == .hourly {
            let baseTime = startDate.timeIntervalSince1970
            let fromTime = max(fromDate.timeIntervalSince1970, baseTime)
            let intervalSeconds = Double(interval * 60 * 60)
            let steps = ceil((fromTime - baseTime) / intervalSeconds)
            let normalizedSteps = max(0, Int(steps))
            return Date(timeIntervalSince1970: baseTime + Double(normalizedSteps) * intervalSeconds)
        }

        if recurring.rule.type == .daily {
            let hour = recurring.rule.time_of_day?.hour ?? Calendar.current.component(.hour, from: startDate)
            let minute = recurring.rule.time_of_day?.minute ?? Calendar.current.component(.minute, from: startDate)

            var candidate = withTime(startDate, hour: hour, minute: minute)
            if candidate < fromDate {
                let deltaDays = Int(floor(fromDate.timeIntervalSince(candidate) / daySeconds))
                let steps = max(0, deltaDays / interval)
                candidate = candidate.addingTimeInterval(Double(steps * interval) * daySeconds)

                while candidate < fromDate {
                    candidate = candidate.addingTimeInterval(Double(interval) * daySeconds)
                }
            }

            return candidate
        }

        if recurring.rule.type == .weekly {
            let daysOfWeek = recurring.rule.weekly_rule?.days_of_week ?? []
            if daysOfWeek.isEmpty { return nil }

            let allowedDays = Array(Set(daysOfWeek.compactMap { dayIndex[$0] })).sorted()
            let hour = recurring.rule.time_of_day?.hour ?? Calendar.current.component(.hour, from: startDate)
            let minute = recurring.rule.time_of_day?.minute ?? Calendar.current.component(.minute, from: startDate)
            let weekSeconds: TimeInterval = 7 * daySeconds
            let anchorWeekStart = startOfWeek(startDate)
            let fromWeekStart = startOfWeek(fromDate)
            let rawWeekDifference = Int(
                floor(fromWeekStart.timeIntervalSince(anchorWeekStart) / weekSeconds)
            )
            let baselineWeekDifference = max(0, rawWeekDifference)
            let remainder = baselineWeekDifference % interval
            var alignedWeekDifference =
                remainder == 0
                ? baselineWeekDifference
                : baselineWeekDifference + (interval - remainder)

            var calendar = Calendar.current
            calendar.timeZone = .current

            for _ in 0..<2 {
                guard let weekStart = calendar.date(
                    byAdding: .day,
                    value: alignedWeekDifference * 7,
                    to: anchorWeekStart
                ) else {
                    alignedWeekDifference += interval
                    continue
                }

                var earliestCandidate: Date?

                for allowedDay in allowedDays {
                    guard let candidateDay = calendar.date(
                        byAdding: .day,
                        value: allowedDay,
                        to: weekStart
                    ) else {
                        continue
                    }

                    let candidate = withTime(candidateDay, hour: hour, minute: minute)
                    if candidate < startDate || candidate < fromDate {
                        continue
                    }

                    if earliestCandidate == nil || candidate < earliestCandidate! {
                        earliestCandidate = candidate
                    }
                }

                if let earliestCandidate {
                    return earliestCandidate
                }

                alignedWeekDifference += interval
            }

            return nil
        }

        return nil
    }

    private func hasPendingForCampaignAt(
        snapshot: CampaignStateSnapshot,
        campaignId: String,
        executeAt: String
    ) -> Bool {
        return snapshot.queued_messages.contains {
            $0.campaign_id == campaignId && $0.execute_at == executeAt
        }
    }

    private func hasFuturePendingForCampaign(
        snapshot: CampaignStateSnapshot,
        campaignId: String,
        now: String
    ) -> Bool {
        let nowDate = parseIsoDate(now)

        for pending in snapshot.queued_messages where pending.campaign_id == campaignId {
            guard let executeAtDate = parseIsoDate(pending.execute_at) else {
                return true
            }

            if let nowDate, executeAtDate >= nowDate {
                return true
            }

            if nowDate == nil {
                return true
            }
        }

        return false
    }

    private func getCampaignState(
        snapshot: CampaignStateSnapshot,
        campaignId: String
    ) -> CampaignStateRecord? {
        return snapshot.campaign_states.first { $0.campaign_id == campaignId }
    }
}
